c++/src/kj/compat/http.h

```diff
@@ -321,7 +321,38 @@ class HttpHeaders {
     kj::StringPtr statusText;
   };
 
-  kj::Maybe<Request> tryParseRequest(kj::ArrayPtr<char> content);
+  struct ProtocolError {
+    // Represents a protocol error, such as a bad request method or invalid headers. Debugging such
+    // errors is difficult without a copy of the data which we tried to parse, but this data is
+    // sensitive, so we can't just lump it into the error description directly. ProtocolError
+    // provides this sensitive data separate from the error description.
+    //
+    // Currently, ProtocolErrors are only reported when parsing requests from clients, and is only
+    // used by HttpServerErrorHandler::handleClientProtocolError().
+    //
+    // TODO(cleanup): Should maybe not live in HttpHeaders? HttpServerErrorHandler::ProtocolError?
+    //   Or HttpProtocolError? Or maybe we need a more general way of attaching sensitive context to
+    //   kj::Exceptions?
+
+    kj::StringPtr description;
+    // An error description safe for all the world to see.
+
+    kj::ArrayPtr<char> rawContent;
+    // Unredacted data which led to the error condition. This may contain anything transported over
+    // HTTP, to include sensitive PII, so you must take care to sanitize this before using it in any
+    // error report that may leak to unprivileged eyes.
+    //
+    // The lifetime of the memory that this ArrayPtr views is the same as the lifetime of the
+    // HttpService::Response object that is passed by reference to
+    // HttpServerErrorHandler::handleClientProtocolError().
+    //
+    // TODO(cleanup): We could make this a self-contained Array<char> for a more intuitive lifetime,
+    //   but this would complicate reporting from within HttpHeaders::tryParseRequest().
+  };
+
+  using RequestOrProtocolError = kj::OneOf<Request, ProtocolError>;
+
+  RequestOrProtocolError tryParseRequest(kj::ArrayPtr<char> content);
   kj::Maybe<Response> tryParseResponse(kj::ArrayPtr<char> content);
```

Comment:
I think if we're updating tryParseRequest() we ought to update tryParseResponse() too. These are technically public interfaces...

Reply:
Done.